docs/proposals/THV-2106-virtual-mcp-server.md

```diff
@@ -0,0 +1,1035 @@
+operational:
+  failure_handling:
+    circuit_breaker:
+      enabled: true
+      failure_threshold: 5       # Open circuit after N failures
+      timeout: 60s               # Stay open for 60s
+```
+
+**States:**
+1. **Closed** (normal): All requests forwarded
+2. **Open**: No requests forwarded (fail immediately)
+3. **Half-Open**: Allow limited requests to test recovery
+
+### Composite Tool Failures
+
+**Per-Step Error Handling:**
+```yaml
+steps:
+  - id: merge
+    tool: github.merge_pr
+    on_error:
+      action: abort  # abort | continue | retry
+      retry_count: 3
+      retry_delay: 5s
+```
+
+**Actions:**
+- `abort`: Stop workflow immediately, return error
+- `continue`: Log error, proceed to next step
+- `retry`: Retry step up to N times with delay
+
+---
+
+## Token Caching Architecture
+
+Virtual MCP implements a pluggable token cache to minimize repeated token exchanges and improve performance.
+
+### Cache Key Structure
+
+Cache keys uniquely identify tokens based on:
+- **Backend name**: Which backend this token is for (e.g., "github")
+- **Subject token hash**: SHA-256 hash of the user's incoming token (for user isolation)
+- **Audience**: Backend audience (e.g., "github-api")
+
+**Key Format:** `{backend}:{hash(subject_token)}:{audience}`
+
+This ensures proper token isolation per (user, backend) pair while maintaining privacy.
+
+### Supported Cache Providers
+
+**Phase 1: Memory Cache**
+- In-memory LRU cache with TTL-based eviction
+- No external dependencies
+- Per-instance cache (not shared across replicas)
+- Suitable for: CLI, single-instance K8s deployments
+
+**Phase 2: Redis Cache**
+- Distributed Redis cache for multi-instance deployments
+- Shared cache across Virtual MCP replicas
+- Persistent token storage (survives restarts)
+- Suitable for: Multi-instance K8s deployments (HA)
+
+**Future: Custom Cache Providers**
+- Plugin architecture for custom cache implementations
+- Register via configuration
+
+### Cache Behavior
+
+**Token Lifetime Management:**
+1. Tokens cached until expiry
+2. Configurable refresh offset (default: 5 minutes before expiry)
+3. Expired tokens automatically evicted
+
+**Performance Characteristics:**
+
+| Operation | Memory Cache | Redis Cache |
+|-----------|-------------|-------------|
+| Cache hit | < 1ms | ~5ms (network) |
+| Cache miss + exchange | 50-200ms | 50-200ms |
+| Memory overhead | ~200 bytes/token | Minimal (client only) |
+| Persistence | No | Yes |
+| Multi-instance | No | Yes |
+
+---
+
+## Implementation Phases
+
+**Phase 1 (MVP)**: Basic aggregation
+- Group-based workload discovery
+- Tool aggregation with filter and override support
+- Simple request routing
+- Pass-through authentication
+
+**Phase 2**: Advanced features
+- Composite tool execution
+- Elicitation support in composite tools
+- Per-backend authentication strategies
+- Token exchange support
+- Failover and load balancing
+
+**Phase 3**: Enterprise features
+- Dynamic configuration updates
```

Comment:
Does audit come into this, or will it piggy back on top of the existing middleware?